pkg/ddl/rollingback.go

```diff
@@ -339,18 +339,184 @@ func convertAddTablePartitionJob2RollbackJob(d *ddlCtx, t *meta.Meta, job *model
+				continue
+			}
+			// Old index marked to be dropped, rollback by making it public again
+			indexInfo.State = model.StatePublic
+			if indexInfo.Global {
+				if id, ok := globalToUniqueDupMap[indexInfo.Name.L]; ok {
+					return ver, errors.NewNoStackErrorf("Duplicate global index names '%s', %d != %d", indexInfo.Name.O, indexInfo.ID, id)
+				}
+				globalToUniqueDupMap[indexInfo.Name.L] = indexInfo.ID
+			}
+		case model.StatePublic:
+			if pi.DDLState != model.StateDeleteReorganization {
+				continue
+			}
+			// We cannot drop the index here, we need to wait until
+			// the next schema version
+			// i.e. rollback in onDropTablePartition
+			// New index that became public in this state,
+			// mark it to be dropped in next schema version
+			if indexInfo.Global {
+				indexInfo.State = model.StateDeleteReorganization
+			} else {
+				// How to know if this index was created as a duplicate or not?
+				if id, ok := globalToUniqueDupMap[indexInfo.Name.L]; ok {
+					// The original index
+					if id >= indexInfo.ID {
+						return ver, errors.NewNoStackErrorf("Indexes in wrong order during rollback, '%s', %d >= %d", indexInfo.Name.O, id, indexInfo.ID)
+					}
+					indexInfo.State = model.StateDeleteReorganization
+				} else {
+					globalToUniqueDupMap[indexInfo.Name.L] = indexInfo.ID
+				}
+			}
+		}
+	}
+	for _, indexInfo := range dropIndices {
+		DropIndexColumnFlag(tblInfo, indexInfo)
+		RemoveDependentHiddenColumns(tblInfo, indexInfo)
+		removeIndexInfo(tblInfo, indexInfo)
+	}
+	if pi.DDLState == model.StateDeleteReorganization {
+		// New partitions are public,
+		// but old is still double written.
+		// OK to revert.
+		// Remove the AddingDefinitions
+		// Add back the DroppingDefinitions
+		if job.Type == model.ActionReorganizePartition {
+			// Reassemble the list of partitions in the same order:
+			// If RANGE, then there should be the same interval
+			// as was replaced!
+			// If KEY/HASH all are replaced
+			// if LIST it needs the following special handling.
+			if pi.Type == model.PartitionTypeList {
+				// Special handling, since the original partitions might not be in order
+				// they should be in the same order within the two arrays.
+				defPos := 0
+				dropPos := 0
+				newDefs := make([]model.PartitionDefinition, 0, len(pi.OriginalPartitionIDsOrder))
+				for _, id := range pi.OriginalPartitionIDsOrder {
+					if defPos < len(pi.Definitions) && pi.Definitions[defPos].ID == id {
+						newDefs = append(newDefs, pi.Definitions[defPos])
+						defPos++
+						continue
+					}
+					if dropPos < len(pi.DroppingDefinitions) && id == pi.DroppingDefinitions[dropPos].ID {
+						newDefs = append(newDefs, pi.DroppingDefinitions[dropPos])
+						dropPos++
+						continue
+					}
+					for {
+						defPos++
+						if defPos < len(pi.Definitions) && pi.Definitions[defPos].ID == id {
+							newDefs = append(newDefs, pi.Definitions[defPos])
+							break
+						}
+					}
```

Comment:
Could you provide a example for this loop, including the value of three slices and the newDefs after this loop? Seems it has some implicit order assumptions for `pi.Definitions` and `pi.DroppingDefinnitions`.

Reply:
Is implicit assumption that reorg partition only affect consecutive partitions?

Reply:
In [`getReplacedPartitionIDs()`](https://github.com/pingcap/tidb/blob/64f5427448b97626660ee65a15fe9fc25086a3a0/pkg/ddl/ddl_api.go#L4566) it will get the first and last index of pi.Definitions that are reorganized (i.e. FROM partitions). If LIST partitioning, the content of `idMap` will be which partitions that are reorganized, and for RANGE it must be a consecutive range of partition (i.e. firstPartIdx to lastPartIdx). HASH and KEY partitioning can only 'ADD' or 'COALESCE' partitions, which means all 'old' partitions will be reorganized to 'new' partitions.
In [`updateDroppingPartitionInfo()`](https://github.com/pingcap/tidb/blob/64f5427448b97626660ee65a15fe9fc25086a3a0/pkg/ddl/partition.go#L1870) it will move the reorganized partitions (old/replaced) from `pi.Definitions` to `pi.DroppingDefinitions` and keeping the original order.
By storing the original order in pi.OriginalPartitionIDsOrder and the property that pi.Definitions and pi.DroppingDefinitions are both ordered according to the original pi.Definitions (before starting the partition reorg), we can do a kind of merge of the pi.Definitions and pi.Dropping definitions and also filter away the 'new' partitions which will not be in the pi.OriginalPartitionIDsOrder array.

So an example would be from the test `TestReorganizePartitionListFailures`:
pi.OriginalPartitionIDsOrder: [135,136,137]
Since there is no given mapping in ALTER TABLE t REORGANIZE PARTITION p0,p2 INTO (... pA ... pB ... pC ...), the new partitions pA,pB,pC will just be placed together where the first reorganized partition's place, in this case p0.
pi.Definitions[].ID: [139,140,141,136]
pi.DroppingDefinitions[].ID: [135,137]
The first loop would not match pi.Definitions, but pi.DroppingDefinitions, so it would get the p0 partition back into its original first position.
newDefs[].ID: [135], defPos: 0, dropPos: 1
The second loop would not directly match pi.Definitions or pi.DroppingDefinitions, so it would skip the three new partitions from pi.Definition (i.e. filter them in the second loop) and then finally find the p1 as the last entry in pi.Definitions.
newDefs[].ID: [135,136], defPos: 3, dropPos: 1
The third loop would not match pi.Definition, but pi.DroppingDefinitions and put back p2 in the original last position in newDefs.
newDefs[].ID: [135,136,137], defPos: 3, dropPos: 2

Notice that the other slices of pi.DroppingDefinitions, pi.AddingDefinitions and pi.Definitions are not changed, but after the loop, pi.Definitions are set to newDefs, which is a new slice.